### PR TITLE
[PT] BatchNorm adaptation

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

Add `nncf.batch_norm_adaptation` function.
Add mode `bn_adaptaion` in example

### Related tickets

174483

### Tests

https://github.com/openvinotoolkit/nncf/actions/runs/19162855749/job/54776694802

### CODE REVIEW

File: examples/pruning/torch/resnet18/main.py [hunk 1]
```diff
@@ -52,7 +52,7 @@ def get_argument_parser() -> ArgumentParser:
     parser.add_argument(
         "--mode",
         type=str,
-        choices=["magnitude", "rb"],
+        choices=["magnitude", "bn_adaptation", "rb"],
```
Comment (daniil-lyakhov) [inline examples/pruning/torch/resnet18/main.py]: Is it possible to give users choice to use bn adaptation regardless of the pruning algorithm? I mean, a `--bn_adaptation` option would be simple IMHO  
Especially because in fact there is a choice: finetune the pruned model or just apply bn adaptation

File: src/nncf/torch/function_hook/pruning/batch_norm_adaptation.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            if num_iterations is not None and idx >= num_iterations:
+                break
+
+            if isinstance(input_data, dict):
+                model(**input_data)
+            elif isinstance(input_data, tuple):
+                model(*input_data)
+            else:
+                model(input_data)
+
```
Comment (daniil-lyakhov) [inline src/nncf/torch/function_hook/pruning/batch_norm_adaptation.py:58]: Can we reuse `PTEngine` here?

File: examples/pruning/torch/resnet18/main.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply examples/pruning/torch/resnet18/main.py]: I’m not sure I see a reason to do that.
What pipeline for bn adaptation you suggest for rb pruning?

File: src/nncf/torch/function_hook/pruning/batch_norm_adaptation.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply src/nncf/torch/function_hook/pruning/batch_norm_adaptation.py:58]: No, in PT Engine used model.eval()

File: src/nncf/pruning/prune_model.py [hunk 3]
```diff
@@ -51,6 +52,30 @@ def prune(
     return model
 
 
+def batch_norm_adaptation(
+    model: TModel, calibration_dataset: Dataset, *, num_iterations: Optional[int] = None
+) -> TModel:
+    """
+    Adapt the batch normalization layers of the given model using the provided dataset.
+    d
```
Comment (ljaljushkin) [inline src/nncf/pruning/prune_model.py]:
```suggestion

```

File: examples/pruning/torch/resnet18/main.py [hunk 4]
```diff
@@ -48,13 +49,18 @@
...
-    parser = ArgumentParser()
+    parser = ArgumentParser(formatter_class=RawTextHelpFormatter)
     parser.add_argument(
         "--mode",
         type=str,
-        choices=["magnitude", "rb"],
-        default="magnitude",
-        help="Pruning mode to use. Choices are: magnitude, rb. Default is magnitude.",
+        choices=["mag", "mag_bn", "rb"],
+        default="mag",
```
Comment (ljaljushkin) [inline examples/pruning/torch/resnet18/main.py:57]: what about testing `bn` together with `mag` in conformance test?

File: src/nncf/pruning/prune_model.py [hunk 5]
```diff
@@ -47,3 +48,27 @@ def prune(
...
+    :param num_iterations: The number of iterations to use for adaptation.
+        If set to None, the adaptation will run for the entire dataset.
+    """
+    backend = get_backend(model)
+    if backend == BackendType.TORCH:
+        from nncf.torch.function_hook.pruning.batch_norm_adaptation import batch_norm_adaptation
+
+        return batch_norm_adaptation(model, calibration_dataset=calibration_dataset, num_iterations=num_iterations)
+
+    msg = f"Pruning is not supported for the {backend} backend."
```
[the remark this answers is not in the document]
Comment (ljaljushkin) [reply src/nncf/pruning/prune_model.py]: Indeed, why pruning?

**Review by ljaljushkin — APPROVED** [review]
